staging/src/k8s.io/apiserver/pkg/authentication/authenticator/metrics.go

```diff
@@ -0,0 +1,195 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package authenticator
+
+import (
+	"context"
+	"net/http"
+	"strings"
+	"time"
+
+	"k8s.io/component-base/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+)
+
+/*
+ * By default, all the following metrics are defined as falling under
+ * ALPHA stability level https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/20190404-kubernetes-control-plane-metrics-stability.md#stability-classes)
+ *
+ * Promoting the stability level of the metric is a responsibility of the component owner, since it
+ * involves explicitly acknowledging support for the metric across multiple releases, in accordance with
+ * the metric stability policy.
+ */
+const (
+	successLabel = "success"
+	failureLabel = "failure"
+	errorLabel   = "error"
+)
+
+var (
+	authenticatedUserCounter = metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Name:           "authenticated_user_requests",
+			Help:           "Counter of authenticated requests broken out by username.",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"username", "authenticator"},
+	)
+
+	authenticatedAttemptsCounter = metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Name:           "authentication_attempts",
+			Help:           "Counter of authenticated attempts.",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"result", "authenticator"},
+	)
+
+	authenticationLatency = metrics.NewHistogramVec(
+		&metrics.HistogramOpts{
+			Name:           "authentication_duration_seconds",
+			Help:           "Authentication duration in seconds broken out by result.",
+			Buckets:        metrics.ExponentialBuckets(0.001, 2, 15),
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"result", "authenticator"},
+	)
+)
+
+// WrapInInstrumentedRequest wraps the input Request in the instrumentedRequest.
+func WrapInInstrumentedRequest(req Request, authenticatorName string) Request {
+	return &instrumentedRequest{authenticatorName: authenticatorName, delegate: req}
+}
+
+type instrumentedRequest struct {
+	// authenticatorName identifies the delegate authenticator.
+	authenticatorName string
+	delegate          Request
+}
+
+// AuthenticateRequest wraps AuthenticateRequest of the delegate Request.
+func (i *instrumentedRequest) AuthenticateRequest(req *http.Request) (*Response, bool, error) {
+	authenticationStart := time.Now()
+	resp, ok, err := i.delegate.AuthenticateRequest(req)
+	recordAuthMetrics(req.Context(), resp, i.authenticatorName, ok, err, authenticationStart)
```

Comment:
Is this going to result in every authenticator logging a deny when a user is rejected?

Reply:
Yes. Is not it what we want?

Reply:
I don't think so?

If a user is authenticated, we want to know which authenticator authenticated them. If the authentication is rejected, I'm not sure how useful it is to know which authenticator rejected them. Unless we could distinguish between "had the right type of credentials, but they were invalid" from "didn't have the credentials".

Reply:
I assume that the case of "had the right type of credentials, but they were invalid" results in !ok, while the case of "didn't have the credentials" results in an error. If that is true, then we do end-up with different labels on the metric: failure vs. error.

I think I need to add more tests to confirm this.

Reply:
I reviewed several authenticators, and it appears that we don't return errors when an incompatible request is received by an authenticator - this makes sense in the context of a union. Instead, we return `nil, false nil` (nil for both: response and error) when the request is not compatible (not what the authenticator expected).
Therefore, I added a condition that will **not** record a metric in such cases. In other words, receiving an incompatible request does not count as processing a request.

However, this implies that when all of the authenticators within a union are incompatible with a request, (as written) no metric will be recorded - this feels wrong since such events are  interesting from monitoring perspective.

I feel that the union package should return an error when none of the authenticators considered the request to be compatible. Though this feels like a significant change in the API. WDYT?